File: genapp-output/nadeemtest/output-example/qt4/nadeemtest/airavata.cpp
#include <glib.h>
#include <iostream>
#include <stdint.h>
#include <sys/time.h>
#include <fstream>

#define _WIN32_WINNT 0x501

#include <thrift/transport/TTransport.h>
#include <thrift/transport/TBufferTransports.cpp>
#include <thrift/transport/TSocket.cpp>
#include <thrift/protocol/TProtocol.h>
#include <thrift/protocol/TBinaryProtocol.h>
#include <thrift/protocol/TBinaryProtocol.tcc>
#include <thrift/TApplicationException.cpp>
#include <thrift/transport/TTransportException.cpp>
#include <thrift/protocol/TProtocolException.h>
#include "../lib/airavata/Airavata.h"
#include "../lib/airavata/Airavata.cpp"
#include "../lib/airavata/airavataDataModel_types.h"
#include "../lib/airavata/airavataDataModel_types.cpp"
#include "../lib/airavata/airavataErrors_types.h"
#include "../lib/airavata/airavataErrors_types.cpp"
#include "../lib/airavata/experimentModel_types.h"
#include "../lib/airavata/experimentModel_types.cpp"
#include "../lib/airavata/workspaceModel_types.h"
#include "../lib/airavata/workspaceModel_types.cpp"
#include "../lib/airavata/airavataAPI_types.h"
#include "../lib/airavata/airavataAPI_types.cpp"
#include "../lib/airavata/applicationDeploymentModel_types.h"
#include "../lib/airavata/applicationDeploymentModel_types.cpp"
#include "../lib/airavata/applicationInterfaceModel_types.h"
#include "../lib/airavata/applicationInterfaceModel_types.cpp"
#include "../lib/airavata/gatewayResourceProfileModel_types.h"
#include "../lib/airavata/gatewayResourceProfileModel_types.cpp"
#include "../lib/airavata/computeResourceModel_types.h"
#include "../lib/airavata/computeResourceModel_types.cpp"


typedef struct {
        gchar *airavata_server, *app_catalog_server;
        gint airavata_port, app_catalog_port, airavata_timeout;
} Settings;

using namespace std;
using namespace apache::thrift;
using namespace apache::thrift::protocol;
using namespace apache::thrift::transport;
using namespace apache::airavata::api;
using namespace apache::airavata::model::workspace::experiment;

void readConfigFile(char* cfgfile, string& airavata_server, int& airavata_port, int& airavata_timeout) {

	airavata_server="'localhost'";
	airavata_port = 8930;
	airavata_timeout = 5000;				

}


string createProject(char* owner, char* projectName)
{
        qDebug() << "IN";   
        int airavata_port, airavata_timeout;
        string airavata_server;
	char* cfgfile;
	cfgfile = "./airavata-client-properties.ini";
        qDebug() << "OUT1";
        readConfigFile(cfgfile, airavata_server, airavata_port, airavata_timeout);	
	qDebug() << "OUT2";			
	airavata_server.erase(0,1);
	airavata_server.erase(airavata_server.length()-1,1);			
	boost::shared_ptr<TSocket> socket(new TSocket(airavata_server, airavata_port));
	socket->setSendTimeout(airavata_timeout);
	boost::shared_ptr<TTransport> transport(new TBufferedTransport(socket));	
	boost::shared_ptr<TProtocol> protocol(new TBinaryProtocol(transport));
	AiravataClient airavataclient(protocol);
	transport->open();
	apache::airavata::model::workspace::Project project;				
	project.owner=owner;
	project.name=projectName;
	string _return;
	airavataclient.createProject(_return,project);				
	transport->close();
        qDebug() << "OUT"; 
	return _return;
				
}

string createExperiment(char* usrName, char* expName, char* projId, char* execId, char* inp)
{
        
        int airavata_port, airavata_timeout;
        string airavata_server;
	char* cfgfile;
	cfgfile = "./airavata-client-properties.ini";
	readConfigFile(cfgfile, airavata_server, airavata_port, airavata_timeout);				
	airavata_server.erase(0,1);
	airavata_server.erase(airavata_server.length()-1,1);			
	boost::shared_ptr<TSocket> socket(new TSocket(airavata_server, airavata_port));
	socket->setSendTimeout(airavata_timeout);
	boost::shared_ptr<TTransport> transport(new TBufferedTransport(socket));	
	boost::shared_ptr<TProtocol> protocol(new TBinaryProtocol(transport));
	AiravataClient airavataclient(protocol);
	transport->open();

	ComputationalResourceScheduling cmRST;
	cmRST.__set_resourceHostId("localhost");
	cmRST.__set_computationalProjectAccount("nadeem");
	cmRST.__set_totalCPUCount(1);
	cmRST.__set_nodeCount(1);
	cmRST.__set_numberOfThreads(0);
	cmRST.__set_queueName("normal");
	cmRST.__set_wallTimeLimit(15);
	cmRST.__set_jobStartTime(0);
	cmRST.__set_totalPhysicalMemory(0);


	UserConfigurationData userConfigurationData;
	userConfigurationData.__set_airavataAutoSchedule(0);
	userConfigurationData.__set_overrideManualScheduledParams(0);
	userConfigurationData.__set_computationalResourceScheduling(cmRST);


	
	char* appId = execId;        

	 
	DataObjectType input;
	input.__set_key("input");
	input.__set_value(inp);
	input.__set_type(DataType::STRING);
	std::vector<DataObjectType> exInputs;
	exInputs.push_back(input);				
	DataObjectType output;
	output.__set_key("output");
	output.__set_value("");
	output.__set_type(DataType::STDOUT);
	std::vector<DataObjectType> exOutputs;
	exOutputs.push_back(output);


	char* user = usrName;
	char* exp_name = expName;
	char* proj = projId;

	Experiment experiment;
	experiment.__set_projectID(proj);
	experiment.__set_userName(user);
	experiment.__set_name(exp_name);
	experiment.__set_applicationId(appId);
	experiment.__set_userConfigurationData(userConfigurationData);
	experiment.__set_experimentInputs(exInputs);
	experiment.__set_experimentOutputs(exOutputs);
				
	string _return = "";
	airavataclient.createExperiment(_return, experiment);

	transport->close();
	return _return;
}

void launchExperiment(char* expId)
{
        
	int airavata_port, airavata_timeout;
	string airavata_server;
	char* cfgfile;
	cfgfile = "./airavata-client-properties.ini";
	readConfigFile(cfgfile, airavata_server, airavata_port, airavata_timeout);				
	airavata_server.erase(0,1);
	airavata_server.erase(airavata_server.length()-1,1);			
	boost::shared_ptr<TSocket> socket(new TSocket(airavata_server, airavata_port));
	socket->setSendTimeout(airavata_timeout);
	boost::shared_ptr<TTransport> transport(new TBufferedTransport(socket));	
	boost::shared_ptr<TProtocol> protocol(new TBinaryProtocol(transport));
	AiravataClient airavataclient(protocol);
	transport->open();						
	airavataclient.launchExperiment(expId, "airavataToken");
	transport->close();				
}

int getExperimentStatus(char* expId)
{
        
	int airavata_port, airavata_timeout;
	string airavata_server;
	char* cfgfile;
	cfgfile = "./airavata-client-properties.ini";
	readConfigFile(cfgfile, airavata_server, airavata_port, airavata_timeout);				
	airavata_server.erase(0,1);
	airavata_server.erase(airavata_server.length()-1,1);			
	boost::shared_ptr<TSocket> socket(new TSocket(airavata_server, airavata_port));
	socket->setSendTimeout(airavata_timeout);
	boost::shared_ptr<TTransport> transport(new TBufferedTransport(socket));	
	boost::shared_ptr<TProtocol> protocol(new TBinaryProtocol(transport));
	AiravataClient airavataclient(protocol);
	transport->open();			
	ExperimentStatus _return;		
	airavataclient.getExperimentStatus(_return, expId);
	transport->close();
	return _return.experimentState;
				
}

string getExperimentOutput(char* expId)
{
        
	int airavata_port, airavata_timeout;
	string airavata_server;
	char* cfgfile;
	cfgfile = "./airavata-client-properties.ini";
	readConfigFile(cfgfile, airavata_server, airavata_port, airavata_timeout);				
	airavata_server.erase(0,1);
	airavata_server.erase(airavata_server.length()-1,1);			
	boost::shared_ptr<TSocket> socket(new TSocket(airavata_server, airavata_port));
	socket->setSendTimeout(airavata_timeout);
	boost::shared_ptr<TTransport> transport(new TBufferedTransport(socket));	
	boost::shared_ptr<TProtocol> protocol(new TBinaryProtocol(transport));
	AiravataClient airavataclient(protocol);
	transport->open();			
	std::vector<DataObjectType> _return;
	airavataclient.getExperimentOutputs(_return, expId);
	transport->close();
	return _return[0].value;
				
}

File: genapp-output/nadeemtest/output-example/html5/airavata.php
<?php

$GLOBALS['THRIFT_ROOT'] = 'lib/Thrift/';
require_once $GLOBALS['THRIFT_ROOT'] . 'Transport/TTransport.php';
require_once $GLOBALS['THRIFT_ROOT'] . 'Transport/TBufferedTransport.php';
require_once $GLOBALS['THRIFT_ROOT'] . 'Transport/TSocket.php';
require_once $GLOBALS['THRIFT_ROOT'] . 'Protocol/TProtocol.php';
require_once $GLOBALS['THRIFT_ROOT'] . 'Protocol/TBinaryProtocol.php';
require_once $GLOBALS['THRIFT_ROOT'] . 'Exception/TException.php';
require_once $GLOBALS['THRIFT_ROOT'] . 'Exception/TApplicationException.php';
require_once $GLOBALS['THRIFT_ROOT'] . 'Exception/TProtocolException.php';
require_once $GLOBALS['THRIFT_ROOT'] . 'Exception/TTransportException.php';
require_once $GLOBALS['THRIFT_ROOT'] . 'Base/TBase.php';
require_once $GLOBALS['THRIFT_ROOT'] . 'Type/TType.php';
require_once $GLOBALS['THRIFT_ROOT'] . 'Type/TMessageType.php';
require_once $GLOBALS['THRIFT_ROOT'] . 'Factory/TStringFuncFactory.php';
require_once $GLOBALS['THRIFT_ROOT'] . 'StringFunc/TStringFunc.php';
require_once $GLOBALS['THRIFT_ROOT'] . 'StringFunc/Core.php';

$GLOBALS['AIRAVATA_ROOT'] = 'lib/Airavata/';
require_once $GLOBALS['AIRAVATA_ROOT'] . 'API/Airavata.php';
require_once $GLOBALS['AIRAVATA_ROOT'] . 'Model/Workspace/Types.php';
require_once $GLOBALS['AIRAVATA_ROOT'] . 'Model/Workspace/Experiment/Types.php';
require_once $GLOBALS['AIRAVATA_ROOT'] . 'API/Error/Types.php';

require_once 'lib/AiravataClientFactory.php';

use Airavata\API\Error\AiravataClientException;
use Airavata\API\Error\AiravataSystemException;
use Airavata\API\Error\InvalidRequestException;
use Airavata\Client\AiravataClientFactory;
use Thrift\Protocol\TBinaryProtocol;
use Thrift\Transport\TBufferedTransport;
use Thrift\Transport\TSocket;
use Airavata\API\AiravataClient;
use Airavata\Model\Workspace\Project;
use Airavata\Model\Workspace\Experiment\Experiment;
use Airavata\Model\Workspace\Experiment\DataObjectType;
use Airavata\Model\Workspace\Experiment\UserConfigurationData;
use Airavata\Model\Workspace\Experiment\ComputationalResourceScheduling;
use Airavata\Model\Workspace\Experiment\DataType;
use Airavata\API\Error\ExperimentNotFoundException;	
use Airavata\Model\Workspace\Experiment\ExperimentState;
use Thrift\Exception\TTransportException;

#airavata functions
function createProject($owner, $projectName){	
	$airavataconfig = parse_ini_file("airavata-client-properties.ini");
	$transport = new TSocket($airavataconfig['AIRAVATA_SERVER'], $airavataconfig['AIRAVATA_PORT']);
	$transport->setSendTimeout($airavataconfig['AIRAVATA_TIMEOUT']);
	$protocol = new TBinaryProtocol($transport);
	$transport->open();
	$airavataclient = new AiravataClient($protocol);
	try
	{	    
		$project = new Project();
		$project->owner = $owner;
		$project->name = $projectName;		
		$projId = $airavataclient->createProject($project);
		
		if ($projId)
		{
		    return "$projId";
		}
		else
		{
		    echo 'Failed to create project.';
		}
	    
	}
	catch (InvalidRequestException $ire)
	{
	    print 'InvalidRequestException: ' . $ire->getMessage()."\n";
	}
	catch (AiravataClientException $ace)
	{
	    print 'Airavata System Exception: ' . $ace->getMessage()."\n";
	}
	catch (AiravataSystemException $ase)
	{
	    print 'Airavata System Exception: ' . $ase->getMessage()."\n";
	}
	$transport->close();
}

function createExperiment($usrName, $expName, $projId, $execId, $inp){	
	$airavataconfig = parse_ini_file("airavata-client-properties.ini");
	$transport = new TSocket($airavataconfig['AIRAVATA_SERVER'], $airavataconfig['AIRAVATA_PORT']);
	$transport->setSendTimeout($airavataconfig['AIRAVATA_TIMEOUT']);
	$protocol = new TBinaryProtocol($transport);
	$transport->open();
	$airavataclient = new AiravataClient($protocol);
	try
	{	    
		/* ComputationalResourceScheduling data for Trestles*/
		$cmRST = new ComputationalResourceScheduling();
		$cmRST->resourceHostId = "localhost";
		$cmRST->ComputationalProjectAccount = "ixxi-2013";
		$cmRST->nodeCount = 1;
		$cmRST->numberOfThreads = 0;
		$cmRST->queueName = "normal";
		$cmRST->wallTimeLimit = 15;
		$cmRST->jobStartTime = 0;
		$cmRST->totalPhysicalMemory = 0;

		$cmRS = $cmRST;
		$userConfigurationData = new UserConfigurationData();
		$userConfigurationData->airavataAutoSchedule = 0;
		$userConfigurationData->overrideManualScheduledParams = 0;
		$userConfigurationData->computationalResourceScheduling = $cmRS;

		
		$appId = $execId;

		/* Experiment input and output data. */
		$input = new DataObjectType();
		$input->key = "input";
		$input->value = $inp;		
		$input->type = DataType::STRING;
		$exInputs = array($input);		
		$output = new DataObjectType();
		$output->key = "output";
		$output->value = "";
		$output->type = DataType::STDOUT;
		$exOutputs = array($output);

		/* Create Experiment: needs to update using unique project ID. */
		$user = $usrName;
		$exp_name = $expName;
		$proj = $projId;

		$experiment = new Experiment();
		$experiment->projectID = $proj;
		$experiment->userName = $user;
		$experiment->name = $exp_name;
		$experiment->applicationId = $appId;
		$experiment->userConfigurationData = $userConfigurationData;
		$experiment->experimentInputs = $exInputs;
		$experiment->experimentOutputs = $exOutputs;

		$expId = $airavataclient->createExperiment($experiment);

		if ($expId)
		{
		    var_dump($experiment);
		    return $expId;
		}
		else
		{
		    return 0;
		}
	}
	catch (InvalidRequestException $ire)
	{
		
	    print 'InvalidRequestException: ' . $ire->getMessage()."\n";
	    return -1;
	}
	catch (AiravataClientException $ace)
	{
	    print 'Airavata System Exception: ' . $ace->getMessage()."\n";
	    return -2;
	}
	catch (AiravataSystemException $ase)
	{
	    print 'Airavata System Exception: ' . $ase->getMessage()."\n";
	    return -3;
	}
	$transport->close();
}

function launchExperiment($expId){
	$airavataconfig = parse_ini_file("airavata-client-properties.ini");
	$transport = new TSocket($airavataconfig['AIRAVATA_SERVER'], $airavataconfig['AIRAVATA_PORT']);
	$transport->setSendTimeout($airavataconfig['AIRAVATA_TIMEOUT']);
	$protocol = new TBinaryProtocol($transport);
	$transport->open();
	$airavataclient = new AiravataClient($protocol);	
	try
	{	   
	   $airavataclient->launchExperiment($expId, 'airavataToken');
	   return 1;
	}
	catch (InvalidRequestException $ire)
	{
	    return -1;
	}
	catch (AiravataClientException $ace)
	{
	    return -2;
	}
	catch (AiravataSystemException $ase)
	{
	    return -3;
	}
	catch (ExperimentNotFoundException $enf)
	{
	    return -4;
	}
	$transport->close();
}

function get_experiment_outputs($expId)
{
    $airavataconfig = parse_ini_file("airavata-client-properties.ini");
	$transport = new TSocket($airavataconfig['AIRAVATA_SERVER'], $airavataconfig['AIRAVATA_PORT']);
	$transport->setSendTimeout($airavataconfig['AIRAVATA_TIMEOUT']);
	$protocol = new TBinaryProtocol($transport);
	$transport->open();
	$airavataclient = new AiravataClient($protocol);

    try
    {
        return $airavataclient->getExperimentOutputs($expId);
    }
    catch (InvalidRequestException $ire)
    {
        echo 'InvalidRequestException!<br><br>' . $ire->getMessage();
    }
    catch (ExperimentNotFoundException $enf)
    {
        echo 'ExperimentNotFoundException!<br><br>' . $enf->getMessage();
    }
    catch (AiravataClientException $ace)
    {
        echo 'AiravataClientException!<br><br>' . $ace->getMessage();
    }
    catch (AiravataSystemException $ase)
    {
        echo 'AiravataSystemException!<br><br>' . $ase->getMessage();
    }
    catch (TTransportException $tte)
    {
        echo 'TTransportException!<br><br>' . $tte->getMessage();
    }
    catch (\Exception $e)
    {
        echo 'Exception!<br><br>' . $e->getMessage();
    }

}


function getOutput($expId){	

	$airavataconfig = parse_ini_file("airavata-client-properties.ini");
	$transport = new TSocket($airavataconfig['AIRAVATA_SERVER'], $airavataconfig['AIRAVATA_PORT']);
	$transport->setSendTimeout($airavataconfig['AIRAVATA_TIMEOUT']);
	$protocol = new TBinaryProtocol($transport);
	$transport->open();
	$airavataclient = new AiravataClient($protocol);
	$outputs = get_experiment_outputs($expId);
	return $outputs;	
	var_dump($outputs);
	$transport->close();
}

function get_experiment_status($expId)
{
    	$airavataconfig = parse_ini_file("airavata-client-properties.ini");
	$transport = new TSocket($airavataconfig['AIRAVATA_SERVER'], $airavataconfig['AIRAVATA_PORT']);
	$transport->setSendTimeout($airavataconfig['AIRAVATA_TIMEOUT']);
	$protocol = new TBinaryProtocol($transport);
	$transport->open();
	$airavataclient = new AiravataClient($protocol);

	    try
	    {
		$experimentStatus = $airavataclient->getExperimentStatus($expId);
	    }
	    catch (InvalidRequestException $ire)
	    {
		echo 'InvalidRequestException!<br><br>' . $ire->getMessage();
	    }
	    catch (ExperimentNotFoundException $enf)
	    {
		echo 'ExperimentNotFoundException!<br><br>' . $enf->getMessage();
	    }
	    catch (AiravataClientException $ace)
	    {
		echo 'AiravataClientException!<br><br>' . $ace->getMessage();
	    }
	    catch (AiravataSystemException $ase)
	    {
		echo 'AiravataSystemException!<br><br>' . $ase->getMessage();
	    }
	    catch (\Exception $e)
	    {
		echo 'Exception!<br><br>' . $e->getMessage();
	    }

	    return ExperimentState::$__names[$experimentStatus->experimentState];
}


function getExperimentStatus($expId){
	$airavataconfig = parse_ini_file("airavata-client-properties.ini");
	$transport = new TSocket($airavataconfig['AIRAVATA_SERVER'], $airavataconfig['AIRAVATA_PORT']);
	$transport->setSendTimeout($airavataconfig['AIRAVATA_TIMEOUT']);
	$protocol = new TBinaryProtocol($transport);
	$transport->open();
	$airavataclient = new AiravataClient($protocol);
	$experimentStatusString = get_experiment_status($expId);
	return $experimentStatusString;
	$transport->close();
}

?>
